### PR TITLE
Feature/login flow

### PR DESCRIPTION
**Pull request by fahademali**

### CODE REVIEW

File: user_service/middlewares/middlewares.go [hunk 1]
```diff
@@ -1 +1,44 @@
...
+
+func NewUserMiddlewares(tokenService services.ITokenService) IUserMiddleware {
+	return &UserMiddleware{
+		tokenService: tokenService,
+	}
+}
+
+func (um *UserMiddleware) Authorize(ctx *gin.Context) {
+	log.GetLog().Info("middleware .............")
+	token := strings.Split(ctx.Request.Header["Authorization"][0], " ")[1]
```
Comment (mvaleed) [inline user_service/middlewares/middlewares.go:30]: The code will break if:
1. No authorization head is passed in headers.
2. Authorization header is passed, but the value is not space separated.

File: user_service/middlewares/middlewares.go [hunk 2]
```diff
@@ -1 +1,44 @@
...
+	user, err := um.tokenService.GetInfoFromToken(token)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, httpResponse.GetErrorResponse(err.Error()))
+		ctx.Abort()
+		return
+	}
+	log.GetLog().Info(user)
+	if user.Role != "admin" {
```
Comment (mvaleed) [inline user_service/middlewares/middlewares.go:38]: 'admin' string should be a constant.
```
UserRoleAdmin = "admin"
```

File: user_service/models/user.go [hunk 3]
```diff
@@ -22,3 +26,10 @@ type User struct {
 	Role       string `db:"role"`
 	IsVerified bool   `db:"is_verified"`
 }
+
+type UserInfo struct {
+	Email      string  `json:"email"`
+	Role       string  `json:"role"`
+	IsVerified bool    `json:"isVerified"`
+	Nbf        float64 `json:"nbf"`
```
Comment (mvaleed) [inline user_service/models/user.go:34]: Whats nbf?

File: user_service/repo/user.go [hunk 4]
```diff
@@ -10,9 +10,9 @@ import (
 )
 
 type IUserRepo interface {
+	SetAdminRole(email string) error
```
Comment (mvaleed) [inline user_service/repo/user.go:13]: UpdateUserByEmail already does this, no?

File: user_service/services/token_service.go [hunk 5]
```diff
@@ -53,3 +58,47 @@ func (ts *TokenService) GetEmailFromAccessToken(accessToken string) (string, err
...
+		if !ok {
+			return models.UserInfo{}, fmt.Errorf("invalid token or missing role claim")
+		}
+
+		return models.UserInfo{
+			Email:      email,
+			IsVerified: isVerified,
+			Nbf:        nbf,
+			Role:       role,
+		}, nil
```
Comment (mvaleed) [inline user_service/services/token_service.go:99]: Avoid unnecessary nested if.
Can be avoided by:
```
claims, ok := token.Claims.(jwt.MapClaims)
If !ok or !token.Valid { return err }
```

File: user_service/middlewares/middlewares.go (hunk 2 above)
Comment (mvaleed) [reply user_service/middlewares/middlewares.go:38]: .

File: user_service/middlewares/middlewares.go [hunk 6]
```diff
@@ -1 +1,44 @@
...
+func (um *UserMiddleware) Authorize(ctx *gin.Context) {
+	log.GetLog().Info("middleware .............")
+	token := strings.Split(ctx.Request.Header["Authorization"][0], " ")[1]
+	user, err := um.tokenService.GetInfoFromToken(token)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, httpResponse.GetErrorResponse(err.Error()))
+		ctx.Abort()
+		return
+	}
+	log.GetLog().Info(user)
```
Comment (mvaleed) [inline user_service/middlewares/middlewares.go:37]: h

**Review by mvaleed — CHANGES_REQUESTED** [review]

File: todo_service/apis/handlers/list.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	"todo_service/services"
+
+	"github.com/gin-gonic/gin"
+)
+
+type IListHandlers interface {
+	HandleCreateList(ctx *gin.Context)
+	HandleDeleteList(ctx *gin.Context)
+	// ASK: shoudl i keep it ListName in below or should i remove list reason of keeping is updaing also means adding a todo.
+	HandleUpdateListName(ctx *gin.Context)
```
Comment (fahademali) [inline todo_service/apis/handlers/list.go]: "Name" not required.

File: todo_service/apis/handlers/list.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		ctx.JSON(http.StatusBadRequest, httpResponse.GetErrorResponse(err.Error()))
+		return
+	}
+
+	ctx.JSON(http.StatusOK, httpResponse.GetSuccessResponse("list has been deleted"))
+}
+
+func (lh *ListHandlers) HandleUpdateListName(ctx *gin.Context) {
+	var requestBody models.UpdateListRequest
+	listID := ctx.Param("listID")
```
Comment (fahademali) [inline todo_service/apis/handlers/list.go]: make listID snake case

File: todo_service/apis/handlers/list.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+	if err := ctx.ShouldBindJSON(&requestBody); err != nil {
+		ctx.JSON(http.StatusBadRequest, err.Error())
+		return
+	}
+
+	if err := lh.listService.UpdateList(listID, requestBody.Name); err != nil {
```
Comment (fahademali) [inline todo_service/apis/handlers/list.go]: send complete request body

File: todo_service/apis/handlers/todo.go [hunk 10]
```diff
@@ -1 +1,96 @@
...
+}
+
+func (th *TodoHandlers) Ping(ctx *gin.Context) {
+	ctx.JSON(http.StatusOK, gin.H{
+		"message": "pong from admin",
+	})
+}
+
+func (th *TodoHandlers) HandleCreateTodo(ctx *gin.Context) {
+	listID := ctx.Param(":listID")
```
Comment (fahademali) [inline todo_service/apis/handlers/todo.go]: make listID snakecase

File: todo_service/apis/handlers/todo.go [hunk 11]
```diff
@@ -1 +1,96 @@
...
+		return
+	}
+
+	ctx.JSON(http.StatusOK, httpResponse.GetSuccessResponse("todo has been created"))
+}
+
+func (th *TodoHandlers) HandleDeleteTodo(ctx *gin.Context) {
+	todoID := ctx.Param("id")
+	//TODO: should i keep todoId as a string
+	if err := th.todoService.DeleteTodo(todoID); err != nil {
```
Comment (fahademali) [inline todo_service/apis/handlers/todo.go]: make it secured

File: todo_service/apis/handlers/todo.go [hunk 12]
```diff
@@ -1 +1,96 @@
...
+		ctx.JSON(http.StatusBadRequest, httpResponse.GetErrorResponse(err.Error()))
+		return
+	}
+
+	ctx.JSON(http.StatusOK, httpResponse.GetSuccessResponse("Todo has been deleted"))
+}
+
+func (th *TodoHandlers) HandleGetTodo(ctx *gin.Context) {
+	todoID := ctx.Param("id")
```
Comment (fahademali) [inline todo_service/apis/handlers/todo.go]: keep it consistent with the struct types

File: todo_service/apis/handlers/todo.go [hunk 13]
```diff
@@ -1 +1,96 @@
...
+	//TODO: should i keep todoId as a string
+	todo, err := th.todoService.GetTodo(todoID)
```
Comment (fahademali) [inline todo_service/apis/handlers/todo.go]: make it secure.

File: todo_service/apis/routes/todo_service.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+	tr.PATCH("/lists/:listID", listHandlers.HandleUpdateListName)
+
+	tr.GET("/ping", todoHandlers.Ping)
+
+	tr.GET("/todos/:id", todoHandlers.HandleGetTodo)
+
+	tr.POST("/todos", todoHandlers.HandleCreateTodo)
+
+	tr.DELETE("/todos/:id", todoHandlers.HandleDeleteTodo)
```
Comment (fahademali) [inline todo_service/apis/routes/todo_service.go]: append list/listid on the route.

File: todo_service/repo/list.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (lr *ListRepo) Delete(id string) error {
+	if _, err := lr.db.Exec("DELETE FROM list where id = $1", id); err != nil {
+		return fmt.Errorf("Delete List Repo: %v", err)
+	}
+	return nil
+}
+
+func (lr *ListRepo) Update(id string, name string) error {
```
Comment (fahademali) [inline todo_service/repo/list.go]: send struct as argument

File: todo_service/services/todo.go [hunk 16]
```diff
@@ -1 +1,86 @@
...
+	err := ts.todoRepo.Insert(requestBody.Title, requestBody.Description, requestBody.DueDate)
+	return err
+}
+
+func (ts *TodoService) DeleteTodo(id string) error {
+	err := ts.todoRepo.Delete(id)
+	return err
+}
+
+func (ts *TodoService) GetTodo(id string) (models.Todo, error) {
```
Comment (fahademali) [inline todo_service/services/todo.go]: clean

File: todo_service/services/todo.go [hunk 17]
```diff
@@ -1 +1,86 @@
...
+	todo, err := ts.todoRepo.Get(id)
+	if err != nil {
+		return models.Todo{}, err
+	}
+	return todo, nil
+}
+
+func (ts *TodoService) UpdateTodo(id string, todoUpdates models.UpdateTodoRequest) error {
```
Comment (fahademali) [inline todo_service/services/todo.go]: make it patch, dont make it put.

File: user_service/apis/handlers/user.go [hunk 18]
```diff
@@ -33,97 +34,100 @@ func (uh *UserHandlers) HandleGetProfile(ctx *gin.Context) {
 	var requestBody models.ProfileRequest
 
 	if err := ctx.ShouldBindJSON(&requestBody); err != nil {
-		ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		errorResponse := httpResponse.GetErrorResponse(err.Error())
+		ctx.JSON(http.StatusBadRequest, errorResponse)
 		return
 	}
-	ctx.JSON(http.StatusOK, gin.H{
-		"data": "/getprofile Not Implemented yet",
```
Comment (fahademali) [inline user_service/apis/handlers/user.go:40]: implement this.

**Review by mvaleed — APPROVED** [review]

**Pull request merged**